[feat] 알림센터 알림 or 푸시알림 클릭 시 BE 로직 구현

seongjunnoh opened the pull request:
## #️⃣ 연관된 이슈

> closes #308

## 📝 작업 내용

1. notifications 테이블 구조 변경
  - 알림 클릭 시 알림에 해당하는 페이지로의 리다이렉트를 위한 데이터들을 저장할 저장소가 필요했습니다
  - 저장할 데이터의 형식이 일정하지 않아 [notifications table에 컬럼 추가, redis 활용, mongo DB 도입] 3가지 방식을 고민하였으나, 기존 MySQL DB + JPA AttributeConverter 를 활용하는 방식이 가장 효율적이라 생각하여 1번 방식으로 구현하였습니다
  - 관련해서 flyway sql 파일 추가했고, 로컬 DB 잘 업데이트 되는것 확인했습니다
<img width="1495" height="149" alt="image" src="https://github.com/user-attachments/assets/ff54d162-b2ad-44f9-9cb8-0e906ee32425" />

2. fcm 푸시알림 구성 수정
  - 기존에는 fcm 서버로 보내는 메시지에 리다이렉트를 위한 데이터까지 모두 포함하였습니다
  - 하지만 '알림 읽음 처리 api' 가 도입되어야 하는 상황에서 리다이렉트를 위한 데이터를 fcm 푸시알림이 아니라, api의 응답으로 포함되는게 더 좋을 것 같다고 생각하여(-> 현재는 웹에서는 푸시알림이 안되므로), 기존 알림 이벤트 퍼블리시, 이벤트 리스닝 코드를 수정하였습니다.
  - 이벤트에는 푸시알림에 필요한 title, content 와 이후 해당 푸시알림의 읽음 처리에 필요한 id에 해당하는 notificationId 값만 포함되도록 수정했습니다.
  - 이 부분은 개발서버에서 BE -> FCM -> FE 로의 통신이 제대로 이루어지는지 확인해봐야할 것 같습니다

3. '알림 읽음 처리 api' 구현
  - 특정 유저가 알림 센터나 푸시 알림을 클릭할 경우 [해당 알림 읽음 처리 + 알림에 해당하는 페이지로 리다이렉트] 되어야 합니다
  - 이 동작을 위한 api 1개를 구현하였습니다
  - 처음에는 FE 플랫폼에 따라 response 를 다르게 구성할까 하다가 (웹은 redirect 할 FE 페이지 url 구성하여 응답, 안드로이드는 기존 redirect data 그대로 응답),FE 플랫폼에 상관없이 [route 경로에 대한 string 값, route 시에 필요한 data] 를 응답하는 것으로 구현했습니다.
  - 이를 응답받아 웹은 각 route 상황에 맞는 페이지 redirect url을 구성하면 되고, 안드로이드는 각 route 상황에 맞는 api 요청 url을 구성하면 될 것 같습니다

## 📸 스크린샷
<img width="587" height="623" alt="image" src="https://github.com/user-attachments/assets/a1ff2663-475e-475b-bf2d-dad3e05a85c4" />
<img width="1262" height="167" alt="image" src="https://github.com/user-attachments/assets/719cdaa3-dc81-4ce8-8ff5-0786f7a587bf" />
-> api 응답 잘 오고, DB의 is_checked 잘 업데이트 되는 것 확인했습니다

## 💬 리뷰 요구사항

### 📌 PR 진행 시 이러한 점들을 참고해 주세요

    * P1 : 꼭 반영해 주세요 (Request Changes) - 이슈가 발생하거나 취약점이 발견되는 케이스 등
    * P2 : 반영을 적극적으로 고려해 주시면 좋을 것 같아요 (Comment)
    * P3 : 이런 방법도 있을 것 같아요~ 등의 사소한 의견입니다 (Chore)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- 신기능
  - 알림 읽음 처리 API 추가: 알림을 읽음으로 표시하고 프론트로 리다이렉트 정보(경로/파라미터) 반환.
- 개선
  - 푸시 페이로드 표준화: 모든 알림에 notificationId 기반 발송·열기 동작 통일.
  - 이벤트/퍼블리셔 페이로드 간소화: 발행 데이터가 notificationId + 대상 사용자 위주로 변경.
  - FCM 관련 요청 검증 메시지 명확화 및 플랫폼 입력 문자열 지원.
- 오류 처리
  - 알림 관련 신규 오류 코드 추가(존재하지 않음, 권한 없음, 이미 읽음, 직렬화/역직렬화 실패, 유효하지 않은 플랫폼).
- 문서화
  - Swagger 응답·오류 케이스 업데이트(알림 읽음 처리, FCM 등록).
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

seongjunnoh committed:
[chore] notifications table 구조 변경을 위한 sql 파일 추가 (#308)

seongjunnoh committed:
[feat] NotificationJpaEntity 에 추가할 data 및 컨버터 추가 (#308)

seongjunnoh committed:
[refactor] 수정된 notification 구조에 따라 jpa, 도메인 엔티티 및 mapper 코드 수정 (#308)

seongjunnoh committed:
[rename] 피드 관련 알림 처리하는 메서드 네이밍 수정 (#308)

seongjunnoh committed:
[refactor] fcm 푸시 알림을 위한 이벤트 퍼블리시 시에 notificationId 추가 및 리다이렉트를 위한 다른 데이터 삭제 (#308)

- 리다이렉트를 위한 데이터들은 fcm 푸시알림에 포함시키는게 아니라, 다른 api 를 통해 응답하도록 수정

seongjunnoh committed:
[refactor] 알림 동기 처리 executor 코드 수정 (#308)

- 알림 DB 저장 시에 RedirectSpec 또한 전달받아 저장하도록 수정
- 저장한 알림의 notificationId 값을 이벤트 퍼블리시 과정에 포함해야하므로, EventCommandInvoker 에게 전달하도록 수정

seongjunnoh committed:
[refactor] FeedNotificationOrchestratorSyncImpl 코드 수정 (#308)

- 알림에 함께 저장해야할 redirectSpec 구성
- 이 값을 포함해서 NotificationSyncExecutor 호출하도록 수정

seongjunnoh committed:
[refactor] FeedNotificationOrchestratorSyncImpl 의 테스트 코드 수정 (#308)

seongjunnoh committed:
[refactor] RoomNotificationOrchestratorSyncImpl 코드 수정 (#308)

- 마찬가지로 redirectSpec 구성 & 이 값을 포함해서 NotificationSyncExecutor 를 호출하도록 수정

seongjunnoh committed:
[refactor] RoomNotificationOrchestratorSyncImpl 의 테스트 코드 수정 (#308)

seongjunnoh committed:
[refactor] fcm 푸시알림을 위한 이벤트 퍼블리시 관련 코드 수정 (#308)

- 퍼블리시하는 이벤트의 구성 수정에 따른 기존 코드 수정

seongjunnoh committed:
[refactor] fcm 푸시알림을 위한 이벤트 리스너 관련 코드 수정 (#308)

- 리스닝하는 이벤트의 구성 수정에 따른 기존 코드 수정

seongjunnoh committed:
[refactor] fcm 푸시알림을 위한 이벤트 리스너 관련 테스트 코드 수정 (#308)

seongjunnoh committed:
[feat] 알림 읽음 처리 api controller 구현 (#308)

seongjunnoh committed:
[feat] 알림 읽음 처리 api use case 구현 (#308)

seongjunnoh committed:
[feat] 알림 읽음 처리 api 영속성 코드 구현 (#308)

seongjunnoh committed:
[feat] 알림 관련 error code 추가 (#308)

seongjunnoh committed:
[feat] 알림 읽음 처리 api 관련 스웨거 ResponseDescription 추가 (#308)

seongjunnoh committed:
[test] 알림 읽음 처리 api 통합 테스트 코드 추가 (#308)

seongjunnoh committed:
[test] Notification 도메인 엔티티 단위 테스트 코드 추가 (#308)

seongjunnoh committed:
[move] PlatformType enum 에 정적 팩토리 메서드 추가 및 패키지 구조 이동 (#308)

- notifications 패키지 -> common/util 패키지로 이동

seongjunnoh committed:
[refactor] 기존 fcm 토큰 등록 api 의 request body 유효성 검증 강화 (#308)

- PlatformType 의 정적 팩토리 메서드를 활용하여 지원하는 FE 플랫폼인지 확인하도록 코드 수정

seongjunnoh committed:
[refactor] PlatformType 패키지 구조 이동에 따른 import 문 수정 (#308)

seongjunnoh committed:
[refactor] MessageRoute 수정 및 패키지 이동 (#308)

- 필요없는 String code value 삭제
- message/domain 패키지 -> notification/domain/value 패키지로 이동

seongjunnoh committed:
Merge remote-tracking branch 'origin' into feat/#308-notification-redirect-data-storage

hd0rable commented on src/main/java/konkuk/thip/feed/application/service/FeedCreateService.java:
```diff
@@ -71,7 +71,7 @@ private void sendNotifications(FeedCreateCommand command, Long savedFeedId) {
         User actorUser = userCommandPort.findById(command.userId());
         for (User targetUser : targetUsers) {
-            feedNotificationOrchestrator.notifyFolloweeNewPost(targetUser.getId(), actorUser.getId(), actorUser.getNickname(), savedFeedId);
+            feedNotificationOrchestrator.notifyFolloweeNewFeed(targetUser.getId(), actorUser.getId(), actorUser.getNickname(), savedFeedId);
```
오호 피드 게시물일때 이름을 좀더 명확하게 수정하셨군욧 좋네요

hd0rable commented on src/main/java/konkuk/thip/message/application/port/in/FeedNotificationDispatchUseCase.java:
```diff
@@ -9,7 +9,7 @@ public interface FeedNotificationDispatchUseCase {
     void handleFeedCommentReplied(FeedEvents.FeedCommentRepliedEvent e);
 
-    void handleFolloweeNewPost(FeedEvents.FolloweeNewPostEvent e);
+    void handleFolloweeNewFeed(FeedEvents.FolloweeNewFeedEvent e);
```
코래가 말한것처럼 위의 3부분도 통일성있게 FolloweeNewFeedEvent로 수정하면 좋을것같습니다!!~

hd0rable commented on src/main/java/konkuk/thip/message/application/service/FeedNotificationDispatchService.java:
```diff
@@ -26,96 +25,66 @@ public class FeedNotificationDispatchService implements FeedNotificationDispatch
     }
 
-    private void pushFeedDetail(Long userId, Notification notification, Long feedId) {
+    private void push(Long userId, Notification n, Long notificationId) {
```
LGTM

hd0rable commented on src/main/java/konkuk/thip/notification/adapter/in/web/request/FcmTokenRegisterRequest.java:
```diff
@@ -18,13 +18,13 @@ public record FcmTokenRegisterRequest(
 
         @NotNull
```
p3: String으로 바뀌었으니 @NotBlank어노테이션으로 바꾸는것이 더 안전할것아 보입니다!!

hd0rable commented on src/main/java/konkuk/thip/notification/domain/value/MessageRoute.java:
```diff
@@ -0,0 +1,22 @@
+
+@Getter
+@RequiredArgsConstructor
+public enum MessageRoute {
```
이부분은 프엔분들과 공유되어야하는 내용이져?? 이넘도 명확하게 네미잉되어있어서 넘 좋습니닷!!

seongjunnoh replied:
넵! String 으로 응답하면 웹, 안드로이드 FE 분들이 각 플랫폼에 따라 redirect 를 위한 후속 작업을 진행해주시는 걸로 얘기나눴습니다!

hd0rable commented on src/main/java/konkuk/thip/notification/application/port/in/NotificationMarkUseCase.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface NotificationMarkUseCase {
+
+    NotificationMarkToCheckedResponse markToChecked(Long notificationId, Long userId);
```
엇 커맨드인데 Response dto 바로사용하셧네요 혹시 다른 이유가 있을까요??

buzz0331 replied:
아마 response의 변수가 두개 이상이라 그런 것 같은데 제 생각엔 읽음 표시가 command의 역할을 하지만 query의 성격을 더 띄고 있는 api라는 생각이 들기는 하는데 다른 분들 생각은 어떠신가요?? (사실 애매해서 큰 상관은 없을 것 같긴합니다)

seongjunnoh replied:
이 부분은 어플리케이션 패키지 전용 응답 모델을 하나 더 추가할까 하다가,
- 추가하는 dto와 response dto의 구성이 동일할 것 같아서 복잡도만 높아질 것 같음
- 조회 로직과 같이 서비스가 response dto 정도는 의존하는 유연한 구조로 가도 문제 없을 것 같음
위 이유로 일단 use case 가 response dto 를 의존하도록 했습니다.

현준님 리뷰처럼 query 에 넣을까도 잠깐 생각해봤는데, 해당 api 가 POST 성격을 띄는 api 라고 생각해서 command 라인의 패키지에 위치시켰습니다

@buzz0331 @hd0rable @coderabbitai

hd0rable commented on src/main/resources/db/migration/V250921__Add_notification_redirect_spec.sql:
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE `notifications`
+  ADD COLUMN `redirect_spec` TEXT NULL
+  COMMENT 'NotificationRedirectSpec을 json 형식의 TEXT 로 저장';
```
붐업~

hd0rable commented on src/test/java/konkuk/thip/notification/adapter/in/web/NotificationMarkToCheckedApiTest.java:
```diff
@@ -0,0 +1,123 @@
+@Transactional
+@AutoConfigureMockMvc(addFilters = false)
+@DisplayName("[통합] 알림 읽음 처리 api 통합 테스트")
+class NotificationMarkToCheckedApiTest {
```
LGTM

hd0rable reviewed (commented):
수고하셨습니다~~ 코드 넘 깔끔하니 좋습니닷!! 👍🏻💯

buzz0331 commented on src/main/java/konkuk/thip/message/application/port/out/RoomEventCommandPort.java:
```diff
@@ -4,53 +4,46 @@ public interface RoomEventCommandPort {
+            Long targetUserId);
 
     // 내가 참여한 모임방 활동이 시작된 경우 (방이 시작 기간이 되어 자동으로 시작된 경우)
     void publishRoomActivityStartedEvent(
-            String title, String content,
-            Long targetUserId, Long roomId, String roomTitle);
+            String title, String content, Long notificationId,
+            Long targetUserId);
 
     // 내가 방장일 때, 새로운 사용자가 모임방 참여를 한 경우
     void publishRoomJoinEventToHost(
-            String title, String content,
-            Long hostUserId, Long roomId, String roomTitle,
-            Long actorUserId, String actorUsername);
+            String title, String content, Long notificationId,
+            Long targetUserId);
 
     // 내가 참여한 모임방의 나의 댓글이 좋아요를 받는 경우
     void publishRoomCommentLikedEvent(
-            String title, String content,
-            Long targetUserId, Long actorUserId, String actorUsername,
-            Long roomId, Integer page, Long postId, String postType);
+            String title, String content, Long notificationId,
+            Long targetUserId);
 
     // 내가 참여한 모임방 안의 나의 기록/투표가 좋아요를 받는 경우
     void publishRoomPostLikedEvent(
-            String title, String content,
-            Long targetUserId, Long actorUserId, String actorUsername,
-            Long roomId, Integer page, Long postId, String postType);
+            String title, String content, Long notificationId,
+            Long targetUserId);
 
     // 내가 참여한 모임방의 나의 댓글에 대댓글이 달린 경우
     void publishRoomPostCommentRepliedEvent(
-            String title, String content,
-            Long targetUserId, Long actorUserId, String actorUsername,
-            Long roomId, Integer page, Long postId, String postType);
+            String title, String content, Long notificationId,
+            Long targetUserId);
```
푸쉬알림으로 화면을 이동할 때 필요한 정보를 message에 담지않고 notificationId만 주입해서 알림센터 조회와 일관성을 맞췄군요! 좋습니다! 일관성이 유지되어서 유지보수가 더 쉬울 것 같아욥

buzz0331 commented on src/main/java/konkuk/thip/notification/adapter/in/web/NotificationCommandController.java:
```diff
@@ -60,4 +64,16 @@ public BaseResponse<Void> deleteFcmToken(
         return BaseResponse.ok(null);
     }
+
+    @Operation(
+            summary = "유저의 특정 알림 읽음 처리",
+            description = "유저가 특정 알림을 읽음 처리합니다. 읽음 처리 후, 해당 알림의 페이지로 리다이렉트를 위한 데이터를 응답합니다."
+    )
+    @ExceptionDescription(NOTIFICATION_MARK_TO_CHECKED)
+    @PostMapping("/notifications/check")
+    public BaseResponse<NotificationMarkToCheckedResponse> markNotificationToChecked(
+            @RequestBody @Valid NotificationMarkToCheckedRequest request,
+            @Parameter(hidden = true) @UserId final Long userId) {
+        return BaseResponse.ok(notificationMarkUseCase.markToChecked(request.notificationId(), userId));
+    }
 }
```
오 읽음 처리랑 리다이렉트 데이터 반환을 통합했군요. 푸쉬알림을 눌렀을 때도 결국 사용자가 알림을 읽은 것이니 푸쉬알림과 알림센터의 알림을 하나의 알림으로 볼 수 있겠네요! 너무 좋습니다

buzz0331 commented on src/main/java/konkuk/thip/common/util/PlatformType.java:
```diff
@@ -0,0 +1,25 @@
+        }
+        throw new InvalidStateException(INVALID_FE_PLATFORM);
+    }
+}
```
음 platformtype은 fcmtoken 테이블에 정의되어 있는 VO라고 생각해서 notification 패키지 하위에 두었던 것 같은데 패키지를 이동시킨 다른 이유가 있을까요??

seongjunnoh replied:
아 FcmTokenJpaEntity 의 컬럼의 타입으로 사용되는 거였군요!!
저는 FE의 platformType 은 전체 프로젝트의 util 적인 성격이 있다고 생각해서 이번 코드 수정시 패키지 이동을 해봤는데, 이러면 notification 패키지 하위의 domain/value 하위로 다시 이동시키겠습니다!!

buzz0331 commented on src/main/java/konkuk/thip/notification/adapter/out/jpa/NotificationJpaEntity.java:
```diff
@@ -34,4 +37,12 @@ public class NotificationJpaEntity extends BaseJpaEntity {
-}
\ No newline at end of file
+
+    @Convert(converter = NotificationRedirectSpecConverter.class)
+    @Column(name = "redirect_spec", columnDefinition = "TEXT")  // nullable
+    private NotificationRedirectSpec redirectSpec;
```
AttributeConverter 활용 좋슴다~

buzz0331 commented on src/main/java/konkuk/thip/notification/application/service/FeedNotificationOrchestratorSyncImpl.java:
```diff
@@ -27,12 +31,19 @@ public class FeedNotificationOrchestratorSyncImpl implements FeedNotificationOrc
     @Transactional(propagation = Propagation.MANDATORY)
     public void notifyFollowed(Long targetUserId, Long actorUserId, String actorUsername) {
         var args = new FollowedTemplate.Args(actorUsername);
+
+        NotificationRedirectSpec redirectSpec = new NotificationRedirectSpec(
+                MessageRoute.FEED_USER,
+                Map.of("userId", actorUserId)
+        );
+
```
오호 Map<String, Object>로 선언해서 동적으로 필요한 리다이렉트 변수를 주입해주었군요! 좋습니다.
다만, 추후에 FE의 요청에 따라 변수명이 바뀔 수도 있는 경우를 대비해서 RedirectDataKey라는 ENUM으로 따로 변수명을 빼두는 것 어떨까요??

seongjunnoh replied:
음 그러면 NotificationRedirectSpec value의 구성을
[MessageRoute, Map<RedirectDataKey Enum, Object>] 이렇게 수정하고, RedirectDataKey Enum 의 value 값을 response 에 포함하는 것을 의미하신 걸까요??

제가 제대로 이해한게 맞는지 여쭤봅니다!

buzz0331 replied:
그렇게 해도 괜찮긴 하지만, 제가 말씀드린건 그냥 USER_ID("userId") 이런식으로 한곳에 모든 key를 선언해두고, 

```java
 NotificationRedirectSpec redirectSpec = new NotificationRedirectSpec(
                MessageRoute.FEED_USER,
                Map.of(RedirectDataKey.USER_ID.getValue(), actorUserId)
  );
```

이런식으로 주입하는 것을 말씀 드린거긴 합니다! 하드코딩되어 있는 변수를 한 곳에 관리하면 유지보수에 더 나을 것 같다고 생각했습니다.

buzz0331 commented on src/main/java/konkuk/thip/notification/application/service/RoomNotificationOrchestratorSyncImpl.java:
```diff
@@ -28,12 +32,24 @@ public class RoomNotificationOrchestratorSyncImpl implements RoomNotificationOrc
                                         Long roomId, Integer page, Long postId, String postType) {
         var args = new RoomPostCommentedTemplate.Args(actorUsername);
+
+        NotificationRedirectSpec redirectSpec = new NotificationRedirectSpec(
+                MessageRoute.ROOM_POST_DETAIL,
+                Map.of(
+                        "roomId", roomId,
+                        "page", page,
+                        "postId", postId,
+                        "postType", postType
+                )
+        );
```
음 코래가 말한 상황을 막기 위해서는 MapUtils 같은 것을 도입해서 Map을 만들때 파라미터가 null-safe인지 확인해서 최종적으로 Map을 만드는 방법이 있을 것 같은데 저는 솔직히 굳이 싶긴하네요.

null-safe를 보장하고 싶으면 고려해봐도 좋을 것 같습니다. (성준님 편하실대로 하셔도 좋다는 뜻)
```java
import java.util.HashMap;
import java.util.Map;

public class MapUtils {

    @SafeVarargs
    public static <K, V> Map<K, V> ofNullable(Object... keyValues) {
        if (keyValues.length % 2 != 0) {
            throw new InvalidStateException(INVALID_MAP_CONSTRUCT);
        }

        Map<K, V> map = new HashMap<>();
        for (int i = 0; i < keyValues.length; i += 2) {
            K key = (K) keyValues[i];
            V value = (V) keyValues[i + 1];
            if (value != null) {
                map.put(key, value);
            }
        }
        return Map.copyOf(map); // 불변 Map 반환
    }
}
```
```java
NotificationRedirectSpec redirectSpec = new NotificationRedirectSpec(
        MessageRoute.ROOM_POST_DETAIL,
        MapUtils.ofNullable(
                "roomId", roomId,
                "page", page,
                "postId", postId,
                "postType", postType
        )
);
```

seongjunnoh replied:
이 부분은 각 알림 생성 대상인 비즈니스 로직에서 이미 검증된 값이라 생각해서, NotificationOrchestrator 구현체에서 전달받은 인자값을 그대로 사용한 거긴 합니다!!

기존 코드를 수정하는 과정에 집중해서 Map.of() 메서드의 NPE 관련해서는 신경쓰지 못했는데, 음 일단 현재 코드 구조로 가보겠습니다!

buzz0331 replied:
네넵 좋슴다

buzz0331 commented on src/test/java/konkuk/thip/notification/adapter/in/web/NotificationMarkToCheckedApiTest.java:
```diff
@@ -0,0 +1,123 @@
+        //given
+        UserJpaEntity user = userJpaRepository.save(TestEntityFactory.createUser(Alias.WRITER));
+
+        NotificationRedirectSpec redirectSpec = TestEntityFactory.createNotificationRedirectSpec(
+                MessageRoute.FEED_USER, Map.of("userId", 123L)    // 특정 유저의 피드 페이지로 이동
+        );
```
Redirect Key를 Enum으로 빼면 이 부분도 모두 바꿀수 있을 것 같네욥

buzz0331 reviewed (commented):
굿굿 좋습니다! 푸쉬알림의 메시지로 NotificationId를 넘겨줌으로써 푸쉬알림과 알림센터의 알림을 하나의 알림으로 봐서 일관성이 지켜진다는 점이 베리 나이스네요~

seongjunnoh committed:
[rename] 네이밍 수정 (#308)

- FolloweeNewPost -> FolloweeNewFeed 로 네이밍 수정

seongjunnoh committed:
[refactor] request dto bean validation error message 추가 (#308)

seongjunnoh commented on src/main/java/konkuk/thip/notification/application/service/RoomNotificationOrchestratorSyncImpl.java:
```diff
@@ -57,12 +85,24 @@ public void notifyRoomVoteStarted(Long targetUserId, Long roomId, String roomTit
+                        "roomId", roomId,
+                        "page", page,
+                        "postId", postId,
+                        "postType", "RECORD"
+                )
```
@coderabbitai 어떻게 수정하는게 좋을까?

seongjunnoh reviewed (commented):
@buzz0331 @hd0rable
일단은 
1. 네이밍 수정
2. request dto bean validation 추가

만 수정해서 커밋했습니다!

현준님 리뷰에서 언급해주신 NotificationRedirectSpec value 에서의 enum 도입과 NotificationJapEntity 에서 해당 컬럼의 nullable 과 관련해서 NPE 발생가능성에 대한 코드 수정은 현재 pr 머지 이후에 추가로 이슈 생성해도 괜찮을까 여쭤봅니다

현재 구현으로 FE 연동에 문제없어 보이고, 오늘 코드 수정할 시간이 부족해서 추후에 다시 리펙해도 되는지 여쭤봅니다 하하

hd0rable commented on src/main/java/konkuk/thip/notification/adapter/in/web/request/FcmTokenRegisterRequest.java:
```diff
@@ -2,29 +2,28 @@
         String fcmToken,
 
-        @NotNull
+        @NotBlank(message = "플랫폼 타입은 필수입니다.")
```
LGTM

seongjunnoh committed:
[move] PlatformType enum 패키지 이동 (#308)

- notification 패키지 하위로 이동
- 프로젝트 전체의 util 적인 성질보다는 notification 패키지 하위에서 사용되는 value 로 생각하는 것으로 결정됨

seongjunnoh committed:
[docs] api 명세 추가 (#308)

hd0rable reviewed (approved):
👍🏻👍🏻
